File: packages/ios/FormbricksSDK/FormbricksSDK/Formbricks.swift
import Foundation
import Network

/// Formbricks SDK delegate protocol. It contains the main methods to interact with the SDK.
public protocol FormbricksDelegate: AnyObject {
    func onSurveyStarted()
    func onSurveyFinished()
    func onSurveyClosed()
    func onError(_ error: Error)
}

/// The main class of the Formbricks SDK. It contains the main methods to interact with the SDK.
@objc(Formbricks) public class Formbricks: NSObject {
    
    static internal var appUrl: String?
    static internal var environmentId: String?
    static internal var language: String = "default"
    static internal var isInitialized: Bool = false
    
    static internal var userManager: UserManager?
    static internal var presentSurveyManager: PresentSurveyManager?
    static internal var surveyManager: SurveyManager?
    static internal var apiQueue: OperationQueue? = OperationQueue()
    static internal var logger: Logger?
    static internal var service = FormbricksService()
    public static weak var delegate: FormbricksDelegate?
    
    // make this class not instantiatable outside of the SDK
    internal override init() {
        /* 
         This empty initializer prevents external instantiation of the Formbricks class.
         All methods are static and the class serves as a namespace for the SDK,
         so instance creation is not needed and should be restricted.
        */
    }
    
    /**
     Initializes the Formbricks SDK with the given config ``FormbricksConfig``.
     This method is mandatory to be called, and should be only once per application lifecycle.
          
     Example:
     ```swift
     let config = FormbricksConfig.Builder(appUrl: "APP_URL_HERE", environmentId: "TOKEN_HERE")
        .setUserId("USER_ID_HERE")
        .setLogLevel(.debug)
        .build()
      
     Formbricks.setup(with: config)
     ```
     */
    @objc public static func setup(with config: FormbricksConfig, force: Bool = false) {
        logger = Logger()
        apiQueue = OperationQueue()
        
        if force {
            isInitialized = false
        }
        
        guard !isInitialized else {
            let error = FormbricksSDKError(type: .sdkIsAlreadyInitialized)
            delegate?.onError(error)
            Formbricks.logger?.error(error.message)
            return
        }
        
        self.appUrl = config.appUrl
        self.environmentId = config.environmentId
        self.logger?.logLevel = config.logLevel
        
        userManager = UserManager()
        if let userId = config.userId {
            userManager?.set(userId: userId)
        }
        if let attributes = config.attributes, !attributes.isEmpty {
            userManager?.set(attributes: attributes)
        }
        if let language = config.attributes?["language"] {
            userManager?.set(language: language)
            self.language = language
        }
    
        presentSurveyManager = PresentSurveyManager()
        surveyManager = SurveyManager.create(userManager: userManager!, presentSurveyManager: presentSurveyManager!)
        userManager?.surveyManager = surveyManager
        
        surveyManager?.refreshEnvironmentIfNeeded(force: force)
        userManager?.syncUserStateIfNeeded()
        
        self.isInitialized = true
    }
    
    /**
     Sets the user id for the current user with the given `String`.
     The SDK must be initialized before calling this method.
          
     Example:
     ```swift
     Formbricks.setUserId("USER_ID_HERE")
     ```
     */
    @objc public static func setUserId(_ userId: String) {
        guard Formbricks.isInitialized else {
            let error = FormbricksSDKError(type: .sdkIsNotInitialized)
                        delegate?.onError(error)
                        Formbricks.logger?.error(error.message)
            return
        }
        
        if let existing = userManager?.userId, !existing.isEmpty {
            logger?.error("A userId is already set (\"\(existing)\") – please call Formbricks.logout() before setting a new one.")
            return
        }
        
        userManager?.set(userId: userId)
    }
    
    /**
     Adds an attribute for the current user with the given `String` value and `String` key.
     The SDK must be initialized before calling this method.
          
     Example:
     ```swift
     Formbricks.setAttribute("ATTRIBUTE", forKey: "KEY")
     ```
     */
    @objc public static func setAttribute(_ attribute: String, forKey key: String) {
        guard Formbricks.isInitialized else {
            let error = FormbricksSDKError(type: .sdkIsNotInitialized)
            delegate?.onError(error)
            Formbricks.logger?.error(error.message)
            return
        }
        
        userManager?.add(attribute: attribute, forKey: key)
    }
    
    /**
     Sets the user attributes for the current user with the given `Dictionary` of `String` values and `String` keys.
     The SDK must be initialized before calling this method.
          
     Example:
     ```swift
     Formbricks.setAttributes(["KEY", "ATTRIBUTE"])
     ```
     */
    @objc public static func setAttributes(_ attributes: [String : String]) {
        guard Formbricks.isInitialized else {
            let error = FormbricksSDKError(type: .sdkIsNotInitialized)
            delegate?.onError(error)
            Formbricks.logger?.error(error.message)
            return
        }
        
        userManager?.set(attributes: attributes)
    }
    
    /**
     Sets the language for the current user with the given `String`.
     The SDK must be initialized before calling this method.
          
     Example:
     ```swift
     Formbricks.setLanguage("de")
     ```
     */
    @objc public static func setLanguage(_ language: String) {
        guard Formbricks.isInitialized else {
            let error = FormbricksSDKError(type: .sdkIsNotInitialized)
            delegate?.onError(error)
            Formbricks.logger?.error(error.message)
            return
        }
        
        if (Formbricks.language == language) {
            return
        }
        
        Formbricks.language = language
        userManager?.set(language: language)
    }
    
    /**
     Tracks an action with the given `String`. The SDK will process the action and it will present the survey if any of them can be triggered.
     The SDK must be initialized before calling this method.
          
     Example:
     ```swift
     Formbricks.track("button_clicked")
     ```
     */
    @objc public static func track(_ action: String) {
        guard Formbricks.isInitialized else {
            let error = FormbricksSDKError(type: .sdkIsNotInitialized)
            delegate?.onError(error)
            Formbricks.logger?.error(error.message)
            return
        }
        
        Formbricks.isInternetAvailabile { available in
            if available {
                surveyManager?.track(action)
            } else {
                Formbricks.logger?.warning(FormbricksSDKError.init(type: .networkError).message)
            }
        }
        
    }
    
    /**
     Logs out the current user. This will clear the user attributes and the user id.
     The SDK must be initialized before calling this method.
          
     Example:
     ```swift
     Formbricks.logout()
     ```
     */
    @objc public static func logout() {
        guard Formbricks.isInitialized else {
            let error = FormbricksSDKError(type: .sdkIsNotInitialized)
            delegate?.onError(error)
            Formbricks.logger?.error(error.message)
            return
        }

        userManager?.logout()
    }
    
    /**
    Cleans up the SDK. This will clear the user attributes, the user id and the environment state.
    The SDK must be initialized before calling this method.
    If `waitForOperations` is set to `true`, it will wait for all operations to finish before cleaning up.
    If `waitForOperations` is set to `false`, it will clean up immediately.
    You can also provide a completion block that will be called when the cleanup is finished.

    Example:
    ```swift
    Formbricks.cleanup()

    Formbricks.cleanup(waitForOperations: true) {
        // Cleanup completed
    }
    ```
     */
    
    @objc public static func cleanup(waitForOperations: Bool = false, completion: (() -> Void)? = nil) {
        if waitForOperations, let queue = apiQueue {
            DispatchQueue.global(qos: .background).async {
                queue.waitUntilAllOperationsAreFinished()
                performCleanup()
                DispatchQueue.main.async {
                    completion?()
                }
            }
        } else {
            apiQueue?.cancelAllOperations()
            performCleanup()
            completion?()
        }
    }

    private static func performCleanup() {
        userManager?.logout()
        userManager?.cleanupUpdateQueue()
        presentSurveyManager?.dismissView()
        presentSurveyManager = nil
        userManager = nil
        surveyManager = nil
        apiQueue = nil
        isInitialized = false
        appUrl = nil
        environmentId = nil
        logger = nil
        language = "default"
    }
}

// MARK: - Check the network connection -
private extension Formbricks {
    static func isInternetAvailabile(completion: @escaping (Bool) -> Void) {
        let monitor = NWPathMonitor()
        let queue = DispatchQueue.global(qos: .background)
        
        monitor.pathUpdateHandler = { path in
            completion(path.status == .satisfied)
            monitor.cancel()
        }
        
        monitor.start(queue: queue)
    }
}

File: packages/android/formbricksSDK/src/main/java/com/formbricks/formbrickssdk/Formbricks.kt
package com.formbricks.formbrickssdk

import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import androidx.annotation.Keep
import androidx.fragment.app.FragmentManager
import com.formbricks.formbrickssdk.api.FormbricksApi
import com.formbricks.formbrickssdk.helper.FormbricksConfig
import com.formbricks.formbrickssdk.logger.Logger
import com.formbricks.formbrickssdk.manager.SurveyManager
import com.formbricks.formbrickssdk.manager.UserManager
import com.formbricks.formbrickssdk.model.error.SDKError
import com.formbricks.formbrickssdk.webview.FormbricksFragment
import java.io.InputStream

@Keep
interface FormbricksCallback {
    fun onSurveyStarted()
    fun onSurveyFinished()
    fun onSurveyClosed()
    fun onError(error: Exception)
}

@Keep
object Formbricks {
    internal lateinit var applicationContext: Context

    internal lateinit var environmentId: String
    internal lateinit var appUrl: String
    internal var language: String = "default"
    internal var loggingEnabled: Boolean = true
    internal lateinit var certificates: MutableList<InputStream>
    private var fragmentManager: FragmentManager? = null
    internal var isInitialized = false
    var callback: FormbricksCallback? = null

    /**
     * Initializes the Formbricks SDK with the given [Context] config [FormbricksConfig].
     * This method is mandatory to be called, and should be only once per application lifecycle.
     * To show a survey, the SDK needs a [FragmentManager] instance.
     *
     * ```
     * class MainActivity : FragmentActivity() {
     *
     *     override fun onCreate() {
     *         super.onCreate()
     *         val config = FormbricksConfig.Builder("http://localhost:3000","my_environment_id")
     *             .setLoggingEnabled(true)
     *             .setFragmentManager(supportFragmentManager)
     *            .build())
     *         Formbricks.setup(this, config.build())
     *     }
     * }
     * ```
     *
     */
    fun setup(context: Context, config: FormbricksConfig, forceRefresh: Boolean = false) {
        if (isInitialized) {
            val error = SDKError.sdkIsAlreadyInitialized
            callback?.onError(error)
            Logger.e(error)
            return
        }

        applicationContext = context

        appUrl = config.appUrl
        environmentId = config.environmentId
        loggingEnabled = config.loggingEnabled
        fragmentManager = config.fragmentManager
        certificates = config.certificateInputStreams?.toMutableList() ?: mutableListOf()
        config.userId?.let { UserManager.set(it) }
        config.attributes?.let { UserManager.setAttributes(it) }
        config.attributes?.get("language")?.let { UserManager.setLanguage(it) }

        FormbricksApi.initialize()
        SurveyManager.refreshEnvironmentIfNeeded(force = forceRefresh)
        UserManager.syncUserStateIfNeeded()

        isInitialized = true
    }

    /**
     * Sets the user id for the current user with the given [String].
     * The SDK must be initialized before calling this method.
     *
     * ```
     * Formbricks.setUserId("my_user_id")
     * ```
     *
     */
    fun setUserId(userId: String) {
        if (!isInitialized) {
            val error = SDKError.sdkIsNotInitialized
            callback?.onError(error)
            Logger.e(error)
            return

        }
        UserManager.set(userId)
    }

    /**
     * Adds an attribute for the current user with the given [String] value and [String] key.
     * The SDK must be initialized before calling this method.
     *
     * ```
     * Formbricks.setAttribute("my_attribute", "key")
     * ```
     *
     */
    fun setAttribute(attribute: String, key: String) {
        if (!isInitialized) {
            val error = SDKError.sdkIsNotInitialized
            callback?.onError(error)
            Logger.e(error)
            return
        }
        UserManager.addAttribute(attribute, key)
    }

    /**
     * Sets the user attributes for the current user with the given [Map] of [String] values and [String] keys.
     * The SDK must be initialized before calling this method.
     *
     * ```
     * Formbricks.setAttributes(mapOf(Pair("key", "my_attribute")))
     * ```
     *
     */
    fun setAttributes(attributes: Map<String, String>) {
        if (!isInitialized) {
            val error = SDKError.sdkIsNotInitialized
            callback?.onError(error)
            Logger.e(error)
            return
        }
        UserManager.setAttributes(attributes)
    }

    /**
     * Sets the language for the current user with the given [String].
     * The SDK must be initialized before calling this method.
     *
     * ```
     * Formbricks.setLanguage("de")
     * ```
     *
     */
    fun setLanguage(language: String) {
        if (!isInitialized) {
            val error = SDKError.sdkIsNotInitialized
            callback?.onError(error)
            Logger.e(error)
            return
        }
        Formbricks.language = language
        UserManager.setLanguage(language)
    }

    /**
     * Tracks an action with the given [String]. The SDK will process the action and it will present the survey if any of them can be triggered.
     * The SDK must be initialized before calling this method.
     *
     * ```
     * Formbricks.track("button_clicked")
     * ```
     *
     */
    fun track(action: String) {
        if (!isInitialized) {
            val error = SDKError.sdkIsNotInitialized
            callback?.onError(error)
            Logger.e(error)
            return
        }

        if (!isInternetAvailable()) {
            val error = SDKError.connectionIsNotAvailable
            callback?.onError(error)
            Logger.e(error)
            return
        }

        SurveyManager.track(action)
    }

    /**
     * Logs out the current user. This will clear the user attributes and the user id.
     * The SDK must be initialized before calling this method.
     *
     * ```
     * Formbricks.logout()
     * ```
     *
     */
    fun logout() {
        if (!isInitialized) {
            val error = SDKError.sdkIsNotInitialized
            callback?.onError(error)
            Logger.e(error)
            return
        }

        UserManager.logout()
    }

    /**
     * Sets the [FragmentManager] instance. The SDK always needs the actual [FragmentManager] to
     * display surveys, so make sure you update it whenever it changes.
     * The SDK must be initialized before calling this method.
     *
     * ```
     * Formbricks.setFragmentManager(supportFragmentMananger)
     * ```
     *
     */
    fun setFragmentManager(fragmentManager: FragmentManager) {
        this.fragmentManager = fragmentManager
    }

    /// Assembles the survey fragment and presents it
    internal fun showSurvey(id: String) {
        if (fragmentManager == null) {
            val error = SDKError.fragmentManagerIsNotSet
            callback?.onError(error)
            Logger.e(error)
            return
        }

        fragmentManager?.let {
            FormbricksFragment.show(it, surveyId = id)
        }
    }

    /// Checks if the phone has active network connection
    private fun isInternetAvailable(): Boolean {
        val connectivityManager = applicationContext.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val network = connectivityManager.activeNetwork ?: return false
        val capabilities = connectivityManager.getNetworkCapabilities(network) ?: return false
        return capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
    }
}